src/main/java/io/temporal/internal/replay/ReplayWorkflowExecutor.java

```diff
@@ -622,27 +341,76 @@ public void close() {
 
   @Override
   public Optional<Payloads> handleQueryWorkflowTask(
-      PollWorkflowTaskQueueResponseOrBuilder response, WorkflowQuery query) throws Throwable {
+      PollWorkflowTaskQueueResponseOrBuilder workflowTask, WorkflowQuery query) {
     lock.lock();
     try {
       AtomicReference<Optional<Payloads>> result = new AtomicReference<>();
-      handleWorkflowTaskImpl(response, () -> result.set(workflow.query(query)));
+      handleWorkflowTaskImpl(workflowTask);
+      result.set(workflow.query(query));
       return result.get();
     } finally {
       lock.unlock();
     }
   }
 
-  public Consumer<HistoryEvent> getLocalActivityCompletionSink() {
-    return localActivityCompletionSink;
+  private void processLocalActivityRequests(long startTime) {
+    Duration maxProcessingTime = getWorkflowTaskTimeout().multipliedBy(4).dividedBy(5);
+    while (true) {
+      List<ExecuteLocalActivityParameters> laRequests =
+          workflowStateMachines.takeLocalActivityRequests();
+      long timeoutInterval = (long) ((System.currentTimeMillis() - startTime) * 0.5);
+      for (ExecuteLocalActivityParameters laRequest : laRequests) {
+        // TODO(maxim): In the presence of workflow task heartbeat this timeout doesn't make
+        // much sense. I believe we should add ScheduleToStart timeout for the local activities
+        // as well.
+        long processingTime = Math.min(System.currentTimeMillis() - startTime, timeoutInterval);
+        boolean accepted =
+            localActivityTaskPoller.apply(
+                new LocalActivityWorker.Task(
+                    laRequest, localActivityCompletionSink, 10000 /* TODO: Configurable */),
+                Duration.ofMillis(processingTime));
```

Comment:
this represents the amount of time the local activity has to execute?

Reply:
This is the maximum time for local retries. I've made it configurable through LocalActivityOptions